# Data Analysis

Now that we've got clean data, let's start with some basic financial analysis.

First, let's load our CSV file into a DataFrame, covert our dates, set the index, and check for duplicated rows or missing values.

In [ ]:
import pandas as pd

df = pd.read_csv("TSLA_clean.csv")
df["Date"] = pd.to_datetime(df["Date"])
df = df.set_index("Date").sort_index().drop_duplicates()

print("Duplicates:", df.duplicated().sum())
print("Missing:", df.isnull().sum().sum())

# we can use code like this here because we cleaned up those data and took them back to the ISO standard
# and then we save that as slate clean it sort of fixed all

Duplicates: 0
Missing: 0


## Returns

Returns refer to the gain or loss made on an initial investment, often expressed as a percentage. We can use the generic **percentage change** formula here:

$$ (price_{end} - price_{start}) / price_{start} $$

We can apply this to close prices to calculate the simple daily return:

$$ (close price_{today} - close price_{yesterday}) / close price_{yesterday} $$

In [ ]:
jan31_closing = df.loc["2024-01-31", "Close"]
jan30_closing = df.loc["2024-01-30", "Close"]

jan31_return = (jan31_closing - jan30_closing) / jan30_closing
print(f"Return on 31 Jan was {jan31_return:.2%}")

# f string: you can insert variables into them 
#.2%: take two decimal points and express it as a percentage 

Return on 31 Jan was -2.24%


In [4]:
name = "Yudie"

print(f"Hi {name} !!")

Hi Yudie !!


This simple daily return expresses a loss in value of 2.24% from one day to the next. Notice we leave our return in decimal form, but when we output it we use `f-strings` and `:.2%` to display it as a percentage.

If we wanted to use the above approach to calculate daily returns for each day in our data set, it would take a long time. Let's see how we can use pandas `pct_change()` to make this sort of work easy, by applying our percentage change formula one column at a time.

In [ ]:
df.Close.pct_change()
# there is a missing value in first row due to no closing value on the first day

df["Returns"] = df.Close.pct_change() # create a new column called Returns in data frame
df

Close        High         Low        Open    Volume   Returns
Date                                                                          
2015-01-02   14.620667   14.883333   14.217333   14.858000  71466000       NaN
2015-01-05   14.006000   14.433333   13.810667   14.303333  80527500 -0.042041
2015-01-06   14.085333   14.280000   13.614000   14.004000  93928500  0.005664
2015-01-07   14.063333   14.318667   13.985333   14.223333  44526000 -0.001562
2015-01-08   14.041333   14.253333   14.000667   14.187333  51637500 -0.001564
...                ...         ...         ...         ...       ...       ...
2024-12-23  430.600006  434.510010  415.410004   28.586000  72698100  0.022657
2024-12-24  462.279999  462.779999  435.140015  435.899994  59551800  0.073572
2024-12-26  454.130005  465.329987  451.019989  465.160004  76366400 -0.017630
2024-12-27  431.660004  450.000000  426.500000  449.519989  82666800 -0.049479
2024-12-30  417.410004  427.000000  415.750000  419.399994  64941000 -0.033012

[2516 rows x 6 columns]

Notice how the first row in our data has a missing value **NaN** in the new daily return column. This is because our data doesn't have a close price for the day before it!

What to do with this missing value depends on what further analysis we want to do. If we want to carry out simple descriptive statistics like compute the mean, max, or standard deviation,  we can leave our missing value as NaN, because pandas will by default ignore NaNs when calculating these.

In [ ]:
df.Returns.max() * 100 # find the max return day 
df.Returns.idxmax() # find the day which had the max return

1 + df.Returns # convert a return to a growth factor
# return is multiplicative and can convert into growth factor which is also multiplicative over time

((1 + df.Returns).prod() - 1) * 100
# the overall return of Tasla over 10 years would have been

np.float64(2754.9312462864846)

For more complex analyses though, we may want to drop or fill this value. Let's calculate cumulative returns for the period. Instead of comparing a given day with the day before it, cumulative returns compare a given day with the first day of the period, to indicate how our stock has performed since our initial investment.

We generally fill missing daily returns with a 0, which indicates no change with the day before.

Because we're doing cumulative multiplication, we'll add 1 to the closing price to get a growth factor, so we can compound the return over time. After calculating the cumulative product, we subtract 1 to get back to a return.

In [ ]:
df["Cummulative"] = (1 + df.Returns).cumprod() - 1
df

Close        High         Low        Open    Volume  \
Date                                                                   
2015-01-02   14.620667   14.883333   14.217333   14.858000  71466000   
2015-01-05   14.006000   14.433333   13.810667   14.303333  80527500   
2015-01-06   14.085333   14.280000   13.614000   14.004000  93928500   
2015-01-07   14.063333   14.318667   13.985333   14.223333  44526000   
2015-01-08   14.041333   14.253333   14.000667   14.187333  51637500   
...                ...         ...         ...         ...       ...   
2024-12-23  430.600006  434.510010  415.410004   28.586000  72698100   
2024-12-24  462.279999  462.779999  435.140015  435.899994  59551800   
2024-12-26  454.130005  465.329987  451.019989  465.160004  76366400   
2024-12-27  431.660004  450.000000  426.500000  449.519989  82666800   
2024-12-30  417.410004  427.000000  415.750000  419.399994  64941000   

             Returns  Cummulative  
Date                               
2015-01-02       NaN          NaN  
2015-01-05 -0.042041    -0.042041  
2015-01-06  0.005664    -0.036615  
2015-01-07 -0.001562    -0.038120  
2015-01-08 -0.001564    -0.039624  
...              ...          ...  
2024-12-23  0.022657    28.451460  
2024-12-24  0.073572    30.618255  
2024-12-26 -0.017630    30.060826  
2024-12-27 -0.049479    28.523960  
2024-12-30 -0.033012    27.549312  

[2516 rows x 7 columns]

### Exercise: Bull or Bear?

Calculate the daily change in trading volume over the data frame. Then calculate the average change in trading volume over the period. Take the same approach we used for calculating daily returns, but considering `Volume` instead of `Close` price.

Then determine the trend in TSLA's stock for **Q1 2024**:

- Rising volume and increasing price might indicate a **bullish** trend (where the uptrend is backed by strong demand and could continue).

- Rising volume and decreasing price might indicate a **bearish** trend (where the downtrend is backed by strong selling pressure and could continue).

- Falling volume on price increase or decrease often indicate a trend is losing strength. It might suggest that momentum is waning and a price reversal is coming.

In [ ]:
## YOUR CODE GOES HERE

## Surges

Surges in price or trading volume can be helpful indicators for traders. One way to define a surge is as an increase on the day before by an amount higher than some defined threshold. That threshold is often defined as some number of standard deviations above the mean. Let's look at price surges of five standard deviations above the mean.

## Moving Averages

Moving averages are a different kind of indicator, one that smooths out small variations in trading data to give a better picture of the overall trend.

A Simple Moving Average (SMA) is one which averages out a price over a specific period. The average is "moving" because when a new day is considered in the period, the oldest date is discarded.

Moving averages can be *fast*, when they cover a short period, or *slow* when they consider a longer period. The longer the period, the more those small variations are smoothed out.

## Volatility

Volatility looks at the degree of variance in a stock, and can be helpful for determining risk. Periods of high standard deviation indicate higher volatility and may suggest a riskier investment.
